homie-influx/src/main.rs

```diff
@@ -0,0 +1,388 @@
+        let line = line?;
+        if !line.starts_with('#') {
+            let parts: Vec<&str> = line.splitn(2, ':').collect();
+            if parts.len() != 2 {
+                eyre::bail!("Invalid line '{}'", line);
+            }
+            let homie_prefix = parts[0].to_owned();
+            let influxdb_database = parts[1].to_owned();
+            mappings.push(Mapping {
+                homie_prefix,
+                influxdb_database,
+            });
+        }
+    }
+
+    Ok(mappings)
+}
+
+/// Construct a new InfluxDB `Client` based on configuration options or defaults.
+///
+/// The database name is not set; make sure to set it before using the client.
+fn get_influxdb_client() -> Result<Client, eyre::Report> {
+    let influxdb_url: Url = std::env::var("INFLUXDB_URL")
+        .unwrap_or_else(|_| DEFAULT_INFLUXDB_URL.to_string())
+        .parse()?;
+    let influxdb_username = std::env::var("INFLUXDB_USERNAME").ok();
+    let influxdb_password = std::env::var("INFLUXDB_PASSWORD").ok();
+
+    // Set empty database name initially; it will be overridden before the client is used.
+    let mut influxdb_client = Client::new(influxdb_url, "");
+    if let (Some(username), Some(password)) = (influxdb_username, influxdb_password) {
+        influxdb_client = influxdb_client.set_authentication(username, password);
+    }
+    Ok(influxdb_client)
+}
+
+/// Construct the `MqttOptions` for connecting to the MQTT broker based on configuration options or
+/// defaults.
+fn get_mqtt_options() -> MqttOptions {
+    let client_name =
+        std::env::var("MQTT_CLIENT_NAME").unwrap_or_else(|_| DEFAULT_MQTT_CLIENT_NAME.to_string());
+
+    let mqtt_host = std::env::var("MQTT_HOST").unwrap_or_else(|_| DEFAULT_MQTT_HOST.to_string());
+
+    let mqtt_port = std::env::var("MQTT_PORT")
+        .ok()
+        .and_then(|val| val.parse::<u16>().ok())
+        .unwrap_or(DEFAULT_MQTT_PORT);
+
+    let mut mqtt_options = MqttOptions::new(client_name, mqtt_host, mqtt_port);
+    mqtt_options.set_keep_alive(5);
+
+    let mqtt_username = std::env::var("MQTT_USERNAME").ok();
+    let mqtt_password = std::env::var("MQTT_PASSWORD").ok();
+    if let (Some(username), Some(password)) = (mqtt_username, mqtt_password) {
+        mqtt_options.set_credentials(username, password);
+    }
+
+    if std::env::var("MQTT_USE_TLS").is_ok() {
+        let mut client_config = ClientConfig::new();
+        client_config.root_store = rustls_native_certs::load_native_certs()
+            .expect("Failed to load platform certificates.");
+        mqtt_options.set_tls_client_config(Arc::new(client_config));
+    }
+
+    mqtt_options
+}
+
+fn spawn_homie_poll_loop(
+    mut event_loop: HomieEventLoop,
+    controller: Arc<HomieController>,
+    influx_db_client: Client,
+) -> JoinHandle<Result<(), PollError>> {
+    task::spawn(async move {
+        loop {
+            if let Some(event) = controller.poll(&mut event_loop).await? {
```

Comment:
I'm currently getting:
```
Error: Mqtt state: Last pingreq isn't acked

Caused by:
   0: Mqtt state: Last pingreq isn't acked
   1: Last pingreq isn't acked
```
which I'm assuming is happening here.

Do we want to do something like this to allow us to debug what's going on? 
```
diff --git a/homie-influx/src/main.rs b/homie-influx/src/main.rs
index f451b2e..91a701b 100644
--- a/homie-influx/src/main.rs
+++ b/homie-influx/src/main.rs
@@ -1,8 +1,6 @@
 use futures::future::try_join_all;
 use futures::FutureExt;
-use homie_controller::{
-    Datatype, Device, Event, HomieController, HomieEventLoop, Node, PollError, Property,
-};
+use homie_controller::{Datatype, Device, Event, HomieController, HomieEventLoop, Node, Property};
 use influx_db_client::reqwest::Url;
 use influx_db_client::{Client, Point, Precision, Value};
 use rumqttc::MqttOptions;
@@ -146,10 +144,14 @@ fn spawn_homie_poll_loop(
     mut event_loop: HomieEventLoop,
     controller: Arc<HomieController>,
     influx_db_client: Client,
-) -> JoinHandle<Result<(), PollError>> {
+) -> JoinHandle<Result<(), eyre::Report>> {
     task::spawn(async move {
         loop {
-            if let Some(event) = controller.poll(&mut event_loop).await? {
+            if let Some(event) = controller
+                .poll(&mut event_loop)
+                .await
+                .wrap_err_with(|| format!("Failed to poll {:?}", controller))?
+            {
                 match event {
                     Event::PropertyValueChanged {
                         device_id,
```

Reply:
If I do that then I get:
```
Error: Failed to poll HomieController { mqtt_client: AsyncClient { request_tx: Sender { .. }, cancel_tx: Sender { .. } }, base_topic: "homie", devices: Mutex { data: {"mijia-bridge": Device { id: "mijia-bridge", homie_version: "4.0", name: Some("Mijia bridge"), state: Lost, implementation: Some("homie-rs"), nodes: {}, extensions: [Extension { id: "org.homie.legacy-stats", version: "0.1.1", homie_versions: ["4.x"] }, Extension { id: "org.homie.legacy-firmware", version: "0.1.1", homie_versions: ["4.x"] }], local_ip: Some("100.115.92.198"), mac: Some("02:42:19:F6:14:A1"), firmware_name: Some("homie-device"), firmware_version: Some("0.3.0"), stats_interval: Some(60s), stats_uptime: Some(480s), stats_signal: None, stats_cputemp: None, stats_cpuload: None, stats_battery: None, stats_freeheap: None, stats_supply: None }} } }
```

(I was running with a modified version of my mijia-homie .env, and hadn't realised that you added the MQTT_ prefix to a bunch of things. I think that means it was connecting to test.mosquitto.org. Waiting for timeouts here is quite time-consuming, so I haven't debugged it much further than that)

Reply:
I don't think dumping the state of the `HomieController` really adds much here, so just added a static string as context.